### PR TITLE
Add hardcoded_value mechanism

### PR DESCRIPTION
**Pull request by astarche**
### What does this Pull Request accomplish?

Add `hardcoded_value` parameter attribute to use in 3 cases:
* Replace the `pass_null` convention on `reserved` params.
* Hardcode `options` to zero on callback methods to disable calling the callback on the registering thread.
* Pass in zero for the size param for scalar attributes.

### Why should this Pull Request be merged?

Standardizes on shared convention and fixes some corner case issues.

### What testing has been done?

Ran and passed all DAQ tests.

### CODE REVIEW

File: source/tests/system/nidaqmx_driver_api_tests.cpp [hunk 1]
```diff
@@ -922,6 +922,21 @@ TEST_F(NiDAQmxDriverApiTests, SetPreScaledUnits_GetPreScaledUnits_ReturnsAttribu
   EXPECT_EQ(ScaleInt32AttributeValues::SCALE_INT32_UNITS_PRE_SCALED_RPM, response.value());
 }
 
+TEST_F(NiDAQmxDriverApiTests, GetScaledUnitsAsDouble_Fails)
```
Comment (astarche) [inline source/tests/system/nidaqmx_driver_api_tests.cpp:925]: Unfortunately, the trick of using the size=0 as a defensive mechanism does not work as well as I'd hoped. It works for getters using the wrong type, but not for setters.

So I think we will need a different way to handle "wrong type" accesses. Part of that will be enums that are organized by type. We can also generate a attribute-to-type validator to double-check. If we're going to allow for pure "attributes we don't know about", we might leave that behind a toggle. Or allow for plugging into the validator from the config file.

I'm slightly leaning towards keeping the hardcoded-zero-size-param. It's doing some good and provides a nice rule-of-3 third use case for the hardcoded_value mechanism.

Comment (reckenro) [reply source/tests/system/nidaqmx_driver_api_tests.cpp:925]: Ah, I see. I agree though, I'd lean towards keeping it as well. Like you said it's doing some good on its own plus gives some more validity to having the hardcoded_value mechanism.

File: source/tests/system/nidaqmx_driver_api_tests.cpp [hunk 2]
```diff
@@ -922,6 +922,21 @@ TEST_F(NiDAQmxDriverApiTests, SetPreScaledUnits_GetPreScaledUnits_ReturnsAttribu
...
+{
+  const auto SCALE_NAME = std::string("TestScale");
+  auto scale_status = create_lin_scale(SCALE_NAME, 0.5);
+  GetScaleAttributeDoubleResponse response;
+  auto status = get_scale_attribute_double(SCALE_NAME, ScaleAttributes::SCALE_ATTRIBUTE_SCALED_UNITS, response);
+
```
Comment (reckenro) [inline source/tests/system/nidaqmx_driver_api_tests.cpp:931]: You can get rid of this extra blank line.

**Review by reckenro — APPROVED** [review]
